Unaligned access in dap4

After investigating some test failures in netcdf-c on arm due to a Bus Error in ncdump I found GETCOUNTER reads an unsigned long long which may not be properly aligned.

```
#0  0xf7f2e244 in delimitSeq (offsetp=<synthetic pointer>, vlentype=0xf7f2e2b8 <delimitSeqArray+208>, compiler=<optimized out>) at d4fix.c:388
        ret = 0
        i = <optimized out>
        offset = 0xf6e57016
        recordcount = <optimized out>
        recordtype = <optimized out>
        ret = <optimized out>
        i = <optimized out>
        offset = <optimized out>
        recordcount = <optimized out>
        recordtype = <optimized out>
        __PRETTY_FUNCTION__ = "delimitSeq"
        u = <optimized out>
        src = <optimized out>
#1  delimitSeqArray (compiler=0x7c798, varortype=<optimized out>, offsetp=0x0) at d4fix.c:368
        offset = 0xf6e57016
        i = 2220021216314720
        dimproduct = <optimized out>
        type = <optimized out>
#2  0xf7f2e940 in NCD4_delimit (compiler=compiler@entry=0x7c798, topvar=0x7dd50, offsetp=offsetp@entry=0xfffedf24) at d4fix.c:208
        ret = 0
        offset = <optimized out>
        __PRETTY_FUNCTION__ = "NCD4_delimit"
#3  0xf7f2f02c in NCD4_processdata (meta=0x7c798) at d4data.c:75
        var = <optimized out>
        ret = 0
        i = 0
        toplevel = 0x958c0
        root = 0xf7f2f02c <NCD4_processdata+128>
        offset = 0xf6e57008
#4  0xf7f2f9c0 in NCD4_open (path=<optimized out>, mode=<optimized out>, basepe=basepe@entry=0, chunksizehintp=chunksizehintp@entry=0x0, mpidata=0x0, mpidata@entry=0xf7ec06e0 <nc_open+32>,
    dispatch=dispatch@entry=0xf7fbadc4 <NCD4_dispatch_base>, ncid=65536) at d4file.c:237
        ret = 0
        d4info = 0x64ad8
        value = <optimized out>
        meta = 0x7c798
        nc = 0x64ab8
        __PRETTY_FUNCTION__ = "NCD4_open"
#5  0xf7ec05f4 in NC_open (path0=path0@entry=0x32c08 "[log][dap4]file:///home/iurt/rpmbuild/BUILD/netcdf-c-4.7.3/dap4_test/daptestfiles/test_sequence_2.syn", omode=<optimized out>, basepe=basepe@entry=0,
    chunksizehintp=chunksizehintp@entry=0x0, useparallel=useparallel@entry=0, parameters=parameters@entry=0x0, ncidp=0xfffee118, ncidp@entry=0xfffee110) at dfile.c:2117
        stat = <optimized out>
        ncp = 0x64ab8
        dispatcher = 0xf7fbadc4 <NCD4_dispatch_base>
        inmemory = <optimized out>
        diskless = <optimized out>
        mmap = <optimized out>
        path = 0x64f30 "file:///home/iurt/rpmbuild/BUILD/netcdf-c-4.7.3/dap4_test/daptestfiles/test_sequence_2.syn#mode=dap4&log"
        model = {impl = 6, format = 0}
        newpath = 0x64f30 "file:///home/iurt/rpmbuild/BUILD/netcdf-c-4.7.3/dap4_test/daptestfiles/test_sequence_2.syn#mode=dap4&log"
```

The crash happens when using GETCOUNTER as the offset is not 8 bytes aligned.

The following patch allowed the tests to pass:

```
--- netcdf-c-4.7.4/libdap4/ncd4.h	2020-03-27 15:33:36.000000000 +0000
+++ netcdf-c-4.7.4/libdap4/ncd4.h	2020-09-30 22:34:57.733422171 +0000
@@ -185,9 +185,15 @@
 #define DECR(offset,size) ((void*)(((char*)(offset))-(size)))
 #define DELTA(p1,p2) ((ptrdiff_t)(((char*)(p1))-((char*)(p2))))
 
+inline d4size_t __read_counter(void *p) {
+    COUNTERTYPE result;
+    memcpy(&result, p, sizeof(COUNTERTYPE));
+    return (d4size_t)result;
+}
+
 #undef GETCOUNTER
 #undef SKIPCOUNTER
-#define GETCOUNTER(p) ((d4size_t)*((COUNTERTYPE*)(p)))
+#define GETCOUNTER(p) __read_counter(p)
 #define SKIPCOUNTER(p) {p=INCR(p,COUNTERSIZE);}
 
 #undef PUSH

```

Given that all calls to GETCOUNTER seem to be followed by SKIPCOUNTER they may also be combined into a single function rather than two macros.

I also noticed many, but not all, calls to GETCOUNTER in d4fix.c then call `if(compiler->swap) swapinline64(...);` on the result, so there may also be some endianness bugs.

I am now getting an infinite loop somewhere and one test never finishes :(

Trying some change I noticed a warning: `ncd4.h:192:24: warning: passing argument 1 of 'memcpy' makes pointer from integer without a cast [-Wint-conversion]` which gave the reason, the argument of memcpy should be &v instead of v!

With this it now passes:

`{COUNTERTYPE v; memcpy(&v,p,sizeof(v)); return (d4size_t)v;}`

--- a/libdap4/d4util.c
+++ b/libdap4/d4util.c
@@ -459,3 +459,11 @@ NCD4_errorNC(int code, const int line, const char* file)
 {
     return NCD4_error(code,line,file,nc_strerror(code));
 }
+
+d4size_t
+NCD4_getcounter(void* p)
+{
+    COUNTERTYPE v;
+    memcpy(&v,p,sizeof(v));
+    return (d4size_t)v;
+}

--- a/libdap4/ncd4.h
+++ b/libdap4/ncd4.h
@@ -197,7 +197,17 @@ extern int nc__dap4(void);
 
 #undef GETCOUNTER
 #undef SKIPCOUNTER
+
+/* Unclear which macros are defined for which compilers.
+   see: https://sourceforge.net/p/predef/wiki/Architectures/
+*/
+#if defined(__ARM_ARCH_7__) || defined(__ARM_ARCH_6__)
+static d4size_t
+GETCOUNTER(void* p)
+{COUNTERTYPE v; memcpy(v,p,sizeof(v)); return (d4size_t)v;}
+#else
 #define GETCOUNTER(p) ((d4size_t)*((COUNTERTYPE*)(p)))
+#endif
 #define SKIPCOUNTER(p) {p=INCR(p,COUNTERSIZE);}
 
 #undef PUSH

--- a/libdap4/ncd4.h
+++ b/libdap4/ncd4.h
@@ -202,9 +202,8 @@ extern int nc__dap4(void);
    see: https://sourceforge.net/p/predef/wiki/Architectures/
 */
 #if defined(__arm__) && __ARM_ARCH < 8
-static d4size_t
-GETCOUNTER(void* p)
-{COUNTERTYPE v; memcpy(v,p,sizeof(v)); return (d4size_t)v;}
+EXTERNL d4size_t NCD4_getcounter(void* p);
+#define GETCOUNTER(p) NCD4_getcounter(p)
 #else
 #define GETCOUNTER(p) ((d4size_t)*((COUNTERTYPE*)(p)))
 #endif /*defined(__arm__) && __ARM_ARCH < 8*/

--- a/libdap4/ncd4.h
+++ b/libdap4/ncd4.h
@@ -201,13 +201,14 @@ extern int nc__dap4(void);
 /* Unclear which macros are defined for which compilers.
    see: https://sourceforge.net/p/predef/wiki/Architectures/
 */
-#if defined(__ARM_ARCH_7__) || defined(__ARM_ARCH_6__)
+#if defined(__arm__) && __ARM_ARCH < 8
 static d4size_t
 GETCOUNTER(void* p)
 {COUNTERTYPE v; memcpy(v,p,sizeof(v)); return (d4size_t)v;}
 #else
 #define GETCOUNTER(p) ((d4size_t)*((COUNTERTYPE*)(p)))
-#endif
+#endif /*defined(__arm__) && __ARM_ARCH < 8*/
+
 #define SKIPCOUNTER(p) {p=INCR(p,COUNTERSIZE);}
 
 #undef PUSH

